File: mobile/androidApp/src/main/kotlin/com/langsapp/android/ui/home/HomeScreen.kt
package com.langsapp.android.ui.home

import androidx.compose.runtime.Composable
import com.langsapp.architecture.Action
import com.langsapp.architecture.ActionSender
import com.langsapp.home.HomeState

@Composable
fun HomeScreen(
    actionSender: ActionSender<Action>,
    homeState: HomeState,
) {
    when (homeState) {
        is HomeState.Welcome -> WelcomeScreen(actionSender, homeState)
        is HomeState.Loading -> LoadingScreen()
        is HomeState.Loaded -> LoadedHomeView(actionSender, homeState)
    }
}

File: mobile/iosApp/iosApp/Screens/Home/HomeScreen.swift
//
//  HomeScreen.swift
//  iosApp
//
//  Created by Szymon Klimek on 07.08.24.
//  Copyright © 2024 orgName. All rights reserved.
//

import app
import SwiftUI

struct HomeScreen: View {
    let state: HomeState
    let actionSender: ActionSender
    
    init(state: HomeState, actionSender: ActionSender) {
        self.state = state
        self.actionSender = actionSender
    }
    
    var body: some View {
        switch state {
        case is HomeState.Welcome:
            WelcomeScreen(state: state as! HomeState.Welcome, actionSender: actionSender)
        case is HomeState.Loading:
            LoadingScreen()
        case is HomeState.Loaded:
            LoadedHomeScreen(state: state as! HomeState.Loaded, actionSender: actionSender)
        default:
            fatalError("Unknown home state: \(state.description)")
        }
    }
}
